Deduplicate input pages in SpillableFinalOnlyGroupedAccumulator for distinct aggregates

pgupta2 opened the pull request:
Test plan - Unit Tests

aweisberg commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -689,6 +716,27 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+                groupIdsBlock = new GroupByIdBlock(groupIdsBlock.getGroupCount(), dedupPage.getBlock(0));
+
+                // Remove the groupId block from distinct Page
+                page = dedupPage.extractChannels(IntStream.range(1, dedupPage.getChannelCount()).toArray());
```
Can this be conditional on the amount of deduplication?

Not that serious about trying to push to optimize this in the case where deduping isn't working, but worth raising the question.

aweisberg replied:
Oh this doesn't copy!

aweisberg reviewed (commented):
First quick pass. This doesn't add a test that checks that this both deduplicates, round trips the data correctly, and maybe something to show it accounts for memory correctly. Unless this relies on an existing unit test. Not sure how testable memory footprint is though.

I need to better understand what `MarkDistinctHash` does (and whether it resolves collisions) and when `evaluateIntermediate` is called

aweisberg commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -181,7 +182,30 @@ public GroupedAccumulator createGroupedAccumulator(UpdateMemory updateMemory)
+                            Optional.empty(),
+                            joinCompiler,
+                            () -> {
+                                updateMemory.update();
```
Preserve the comments here? Or maybe don't copy paste it and make it a shared lamba in all 3 instances in the file?

aweisberg commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -689,6 +716,27 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
+                groupIdsBlock = new GroupByIdBlock(groupIdsBlock.getGroupCount(), dedupPage.getBlock(0));
+
+                // Remove the groupId block from distinct Page
+                page = dedupPage.extractChannels(IntStream.range(1, dedupPage.getChannelCount()).toArray());
```
Now my question is which step here actually removes the non-distinct rows from memory?

aweisberg replied:
OK, I wasn't sure if `getPositions` on Block actually always creates a copy or a view of the block. Looking at the implementation in say `RunLengthEncodedBlock` it makes no sense to me because it discards the positions and just returns another wrapper. `DictionaryBlock` also seems to play similar games.

The default implementation creates a DictionaryBlock which I think is just a wrapper around the source block?

aweisberg replied:
I think the key difference is that the filtered stuff is normally fed to the underlying aggregation. So the reduction happens there and most of this is concerned with identifying what to feed to it. But here I think we just store the page away because we won't feed it to the aggregation until much later.

pgupta2 replied:
> Depending on how MarkDistinctHash works is it going to store these values twice?

MarkDistinctHash will store each distinct row only once. It internally uses GroupByHash which has all the logic to detect duplicates. When duplicate row comes in, it computes the hash for this duplicate row and if this hash has been seen before, then it performs an actual check of values if they are same or not. If equality check passes, then it means that it is actually a duplicate row. If equality check does not passes, then it means that there was a hash collision and it is handled accordingly. So, basically, each distinct row is stored only once.

> If there are multiple distinct aggregations on the same column with no filter or the same filter then we would also like to share the memory used to track distinct values.

Yes, that would be good to have. But, I am not targeting this in this PR. Currently, each aggregate has its own MarkDistinctHash that is used to perform deduplication. Maybe, this can be done as an optimization in followup PR.

aweisberg replied:
I probably put out too much noise so the signal got lost here.

`filter` doesn't actually filter. It returns a `Page` that appears to be a view of the original `Page`'s blocks. This is retaining all non-distinct values for the distincted columns. Please correct me if I am wrong, but all the `getPositions` implementations I looked at create views of the original `Block`.

`MarkDistinctHash` copies the column values being distincted so we are storing the values twice. Once in `MarkDistinctHash` and once in the pages that we fed to mark distinct hash, but also retained. We don't need to do that. We could only retain `MarkDistinctHash` and then extract the values stored in it.

pgupta2 replied:
> filter doesn't actually filter. It returns a Page that appears to be a view of the original Page's blocks. This is retaining all non-distinct values for the distincted columns. Please correct me if I am wrong, but all the getPositions implementations I looked at create views of the original Block.

This is correct. `filter` calls `getPositions` which creates a view over the existing block. But, when we prepare pages for spill, we only spill filtered values and during unspill,  create a page with only distinct rows. We can actually call `copyPositions` instead which performs the actual copy and returns a new page which only contains distinct value.  This will incur some cpu cost due to copy. How expensive is copy operation? If we go with copy approach, we will save on memory buffer as only distinct values will be kept and buffered.


> MarkDistinctHash copies the column values being distincted so we are storing the values twice. Once in MarkDistinctHash and once in the pages that we fed to mark distinct hash, but also retained. We don't need to do that. We could only retain MarkDistinctHash and then extract the values stored in it.

This is actually a good observation. We can actually get values directly from MarkDistinctHash but MarkDistinctHash only stores the value for hash channels. It does not store values for maskChannel which is also required in DistinctingGroupedAccumulator in some cases. Its possible to modify markDistinctHash to start storing maskChannel values as well. Let me give it a deeper thought tomorrow.

aweisberg replied:
I think the issue here is that turning on spilling with this implementation causes queries to use more memory thus forcing them to spill and possibly run out of spill space.

To roll out spilling for distinct aggregation for all queries we need to be at least as memory efficient as the non-spilling version. There is no justification for a regression and on the wider workload it really will cause problems.

To target specific queries where it helps and doesn't hurt then sure we can use this approach, but it is pretty wrong to store non-distinct values in memory. That isn't the right behavior for aggregation.

aweisberg reviewed (changes requested):
I have to say on some level it is impressive how little code this ended up being.

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -689,6 +716,27 @@ public void addInput(GroupByIdBlock groupIdsBlock, Page page)
             rawInputs.ensureCapacity(rawInputsLength);
 
+            // Perform dedup for distinct aggregates
+            if (delegate instanceof DistinctingGroupedAccumulator) {
```
Haven't read the logic yet. A quick design nit:
- instead of doing type check, we can introduce a method for `FinalOnlyGroupedAccumulator` interface called `Page addInputPreprocessing(GroupByIdBlock groupIdsBlock, Page page)` with different implementations for `DistinctingGroupedAccumulator` and `OrderingGroupedAccumulator` (which is a no op)
- `markDistinctHash` is better not visible at `SpillableFinalOnlyGroupedAccumulator` but inside `DistinctingGroupedAccumulator`.
- Would be good to clarify the markDistinctHash already in `DistinctingGroupedAccumulator` and the new one you just introduced. Haven't read the code yet. Are they the same?

pgupta2 commented:
> First quick pass. This doesn't add a test that checks that this both deduplicates, round trips the data correctly, and maybe something to show it accounts for memory correctly. Unless this relies on an existing unit test. Not sure how testable memory footprint is though.


TestSpilledAggregations has multiple test cases that executes the spill path for distinct aggregates. The changes are in common code path and it will be executed for all distinct aggregate test cases by default. I am not sure how we can add a unit test that can check for partial states inside HashAggregationOperator. Any pointers will be very helpful. Same with memory footprint. How can I add a unit test for doing memory checks?

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/FinalOnlyGroupedAccumulator.java:
```diff
@@ -42,4 +42,8 @@ public final void evaluateIntermediate(int groupId, BlockBuilder output)
         throw new UnsupportedOperationException();
     }
+
+    abstract GroupIdPage addInputPreprocessing(GroupIdPage groupIdPage);
+
+    abstract void reset();
```
maybe add javadocs to these two methods?

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -629,6 +677,17 @@ public void prepareFinal()
+        }
+
+        @Override
+        void reset()
```
nit, add a comment inside:
```java
void reset()
{
        // no-op
}

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -757,6 +819,10 @@ public void evaluateIntermediate(int groupId, BlockBuilder output)
             if (blockBuilders == null) {
                 checkState(rawInputs != null);
 
+                // Release memory held by markDistinctHash in delegate.
```
let's avoid using the implementation detail comments like "markDistinctHash" in the interface invocation

highker reviewed (commented):
The structure looks great! Will take a look on logic details tomorrow.

pgupta2 commented:
@highker : I discovered a bug during verifier run and it seems that doing partial dedup before applying filter mask can return incorrect results in some cases. This is because dedup results depends on the order of input rows and markDistinctHash does not consider maskChannel while computing hash.

For example, consider the following data 

Custkey     OrderPriority     OrderKey
   2                  HIGH                90
   2                  HIGH                200
   2                  LOW                 50

Query: SELECT custkey, count(DISTINCT orderpriority) FILTER(WHERE orderkey > 100) FROM orders GROUP BY custkey

In addInputPreprocessing, we first perform dedup using markDistinctHash which computes hashes on groupId + input channels. Since, custKey is group by key and orderpriority is input channel, hash will be computed on these two columns. When we will perform dedup in addInputPreprocessing(), the second row from the input will be removed and first and third row will be pushed to delegate. Delegate (DistinctingGroupedAccumulator) will apply filter based on maskChannel (orderKey) and thus will filter out both the rows, thereby producing incorrect result.

There are two ways to fix this:
1. We extend markDistinctHash to consider maskChannel as well while computing hash
2. We apply maskChannel filter first before doing dedup.

Option #2 is better since it eliminate rows that will be eventually get filtered out inside delegate in preprocessing itself, thereby reducing memory usage and spilled bytes. 
I am updating this PR with the fix. Now, in addInputPreprocessing, we first apply filter based on maskChannel and then we perform dedup. I have also added a unit test that checks this specific codepath.

highker commented:
@pgupta2, thanks for the fix. I haven't got a chance to look into logic details yet. Will pay attention to it while reviewing. On the other side, shall we introduce a config + session property to guard this new behavior? It would be useful to prevent production correctness issues.

pettyjamesm commented on presto-common/src/main/java/com/facebook/presto/common/block/ColumnarRow.java:
```diff
@@ -147,6 +148,17 @@ public int getPositionCount()
 
+    public int getNotNullPositionCount()
+    {
+        int count = 0;
```
Nit: if `!nullCheckBlock.mayHaveNull()` you could `return getPositionCount()`

pettyjamesm commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -488,6 +504,46 @@ public void evaluateFinal(int groupId, BlockBuilder output)
+            GroupByIdBlock distinctGroupIdsBlock = new GroupByIdBlock(groupIdsBlock.getGroupCount(), dedupPage.getBlock(0));
+
+            // Remove the groupId block from distinct Page
+            Page distinctPage = dedupPage.extractChannels(IntStream.range(1, dedupPage.getChannelCount()).toArray());
```
Creating `IntStream.range(1, dedupPage.getChannelCount()).toArray()` for each input page seems unnecessarily expensive, I suggesting storing that array as a field to avoid the per call overhead.

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -181,7 +182,7 @@ public GroupedAccumulator createGroupedAccumulator(UpdateMemory updateMemory)
         aggregateInputChannels.addAll(inputChannels);
         maskChannel.ifPresent(aggregateInputChannels::add);
-        return new SpillableFinalOnlyGroupedAccumulator(sourceTypes, aggregateInputChannels.build(), (FinalOnlyGroupedAccumulator) accumulator);
+        return new SpillableFinalOnlyGroupedAccumulator(sourceTypes, aggregateInputChannels.build(), (FinalOnlyGroupedAccumulator) accumulator, isSpillableAccumulatorPreprocessingEnabled(session));
```
Let's checkArg session is not null. It's interesting that session could be null when `distinct` is false by reading the context.

highker commented on presto-main/src/main/java/com/facebook/presto/sql/analyzer/FeaturesConfig.java:
```diff
@@ -129,6 +129,7 @@
     private boolean orderByAggregationSpillEnabled = true;
     private boolean windowSpillEnabled = true;
     private boolean orderBySpillEnabled = true;
+    private boolean spillableAccumulatorPreprocessingEnabled = true;
```
shall we make it to false? It is safer to rollout in prod

pgupta2 replied:
I wanted to keep this enabled in code so that it can go though out strict release verification process. That way, we would be able to detect any regression/issues during release verification itself. If we disable it in code and the enable it via config in prod, then we will miss the exhaustive testing coverage that we would have gotten otherwise.  

I am also going to perform exhaustive verifier runs to test this PR but still any additional testing is cherry on top. Let me know what you think and if you have strong opinion for disabling it.

highker replied:
lol, don't trust our release process. We don't validate spilling during our release ....

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -723,13 +798,16 @@ public void addIntermediate(GroupByIdBlock groupIdsBlock, Block block)
             // expand array block back into page
             ColumnarArray columnarArray = toColumnarArray(block); // flattens the squashed arrays; so there is no need to flatten block again.
             ColumnarRow columnarRow = toColumnarRow(columnarArray.getElementsBlock()); // contains the flattened array
-            int newPositionCount = columnarRow.getPositionCount(); // number of positions in expanded array (since columnarRow is already flattened)
+            int newPositionCount = columnarRow.getNotNullPositionCount(); // number of positions in expanded array (since columnarRow is already flattened)
             long[] newGroupIds = new long[newPositionCount];
             boolean[] nulls = new boolean[newPositionCount];
             int currentRowBlockIndex = 0;
             for (int groupIdPosition = 0; groupIdPosition < groupIdsBlock.getPositionCount(); groupIdPosition++) {
                 for (int unused = 0; unused < arrayBlock.getBlock(groupIdPosition).getPositionCount(); unused++) {
                     // unused because we are expanding all the squashed values for the same group id
+                    if (arrayBlock.getBlock(groupIdPosition).isNull(unused)) {
+                        break;
+                    }
```
Not quite get this newly changed logic. Could we briefly explain?

pgupta2 replied:
While pre-processing, we first perform a filter on maskChannel. This filter can cause some groupIds to get filtered out completely i.e.. we might end up no rows for the given groupId. In evaluateIntermediate(), we get a groupId as an input and we basically store all the rows for the given groupId in its corresponding `output` block. Since, all rows for the given groupId has been filtered, we basically store a null in the `output` BlockBuilder. This is being done here:

```
if (blockBuilders.get(groupId) == null) {
       singleArrayBlockWriter.appendNull();
}
```

Since, we are storing null for certain groupIds, inside addIntermediate, we can expect output block for a groupId to be null. This logic checks for that case and skip the groupIds for which there are no rows.

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -488,6 +504,46 @@ public void evaluateFinal(int groupId, BlockBuilder output)
+            Block distinctMask = work.getResult();
+
+            // Filter out duplicate rows and return the page with distinct rows
+            Page dedupPage = filter(filtered, distinctMask);
```
Likely we could do a `compact` when filtering ratio is high.

aweisberg replied:
I think even if it compacted every time it would still be a 2x memory regression because it is storing the distinct values twice compared to the non-spill case where the values are stored once in `MarkDistinctHash` and 0-1x inside the wrapped accumulator.

pgupta2 replied:
I am currently working on adding support for using MarkDistinctHash as the input source when pre-processing is enabled. I have the changes ready and unit tests are passing but I want to do some verifier runs to ensure that changes are working correctly. With these changes, we will no longer buffer pages in `rawInputs` when pre-processing is enabled. Instead, we will get them from `markDistinctHash` right before spilling to disk or pushing to underlying delegate.

With these changes, I am going to restructure the code in the following way:
1. `SpillableFinalOnlyGroupedAccumulator` will be used when pre-processing is disabled or when delegate is OrderingGroupedAccumulator. Some changes are done in this class to support pre-processed data but these changes are backward compatible with non-processed inputs as well.
2. We will add a new class called `SpillableDistinctPreprocessingFinalOnlyGroupedAccumulator` which extends `SpillableFinalOnlyGroupedAccumulator` and is used only for distinct aggregates when pre-processing is enabled. This will allow us to write distinct specific logic without adding too many methods in `FinalOnlyGroupedAccumulator` interface.
3. We will remove the `addInputPreprocessing()` and `reset()` API from `FinalOnlyGroupedAccumulator` since these applies only to distinct aggregates and could be called directly inside `SpillableDistinctPreprocessingFinalOnlyGroupedAccumulator`
4. `SpillableDistinctPreprocessingFinalOnlyGroupedAccumulator` will override necessary methods like addInput, evaluateIntermediate, prepareFinal and will perform logic that is specific to distinct aggregates with preprocessing enabled.

With these changes, we will be able to cleanly separate out logic when pre-processing is enabled vs disabled. I will update the PR once I have done some verifier testing but would be eager to get feedback on the new design.

highker commented on presto-main/src/main/java/com/facebook/presto/operator/aggregation/GenericAccumulatorFactory.java:
```diff
@@ -488,6 +504,46 @@ public void evaluateFinal(int groupId, BlockBuilder output)
         }
+
+        @Override
+        public GroupIdPage addInputPreprocessing(GroupIdPage groupIdPage)
```
This method has a lot of duplication with `addInput`. Can we improve on that?

pgupta2 commented:
@highker @aweisberg @pettyjamesm : Updated the PR to use markDistinctHash as input source for Dedup pages. This will allows us to skip buffering in `rawInputs` when pre-processing is enabled. This should solve two problems for us:
1. It removes 2x memory usage during pre-processing (Pages were buffered in `rawInput` and then distinct rows are copied inside `markDistinctHash`). Now, we don't buffer pages in `rawInputs`.
2. MarkDistinctHash stores only distinct rows internally. When we get distinct pages from `markDistinctHash`, we get back distinct compacted pages. Previously, the dedup logic would create a view (DictionaryBlock) over the existing page which would not reduce buffering memory usage. With new approach, we are going to actually buffer only dedup rows. This should reduce memory usage.

pgupta2 commented:
> Test failure is related

I noticed the failure but didn't updated the PR since there was a bug that I found and I wanted to include the fix as well in the update. The bug triggers when aggregation maskFilter filters out all the input rows. In that case, nothing is sent to delegate's accumulator and it throws a NLP (since its internal states are not initialized). The fix is to send an empty page in such case with the correct groupCount so that delegate's accumulator can init its internal states. The bug is now fixed and I have added a unit test to test this.

highker commented:
LGTM. @aweisberg, do you have other comments on this PR?
